src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs

```diff
@@ -0,0 +1,365 @@
+                {
+                    collectedTypes.AddStruct(model.GetDeclaredSymbol(@struct));
+                }
+
+                var namespaces = tree.GetRoot().DescendantNodes().OfType<NamespaceDeclarationSyntax>();
+                foreach (var @namespace in namespaces)
+                {
+                    collectedTypes.AddNamespace(model.GetDeclaredSymbol(@namespace));
+                }
+            }
+            return collectedTypes;
+        }
+
+        /// <summary>Make a suggestion for types to use instead of the given type</summary>
+        /// <param name="type">A type that is not valid in Windows Runtime</param>
+        /// <returns>string of types that the given type implements and are valid Windows Runtime types</returns>
+        private string SuggestType(string type)
+        {
+            switch (type)
+            {
+                case "System.Collections.Generic.Dictionary`2": 
+                    return "IDictionary<TKey,TValue>, IReadOnlyDictionary<TKey,TValue>, IEnumerable<KeyValuePair<TKey,TValue>>";
+                case "System.Collections.ObjectModel.ReadOnlyDictionary`2":
+                    return "IReadOnlyDictionary<TKey,TValue>, IEnumerable<KeyValuePair<TKey,TValue>>, IDictionary<TKey,TValue>";
+                case "System.Collections.Generic.List`1": 
+                    return "IList<T>, IReadOnlyList<T>, IEnumerable<T>";
+                case "System.Linq.Enumerable`1": 
+                    return "IEnumerable<T>";
+                case "System.Collections.Generic.KeyValuePair": 
+                    return "KeyValuePair<TKey,TValue>";
+                case "System.Array": 
+                    return "T[]";
+                default: return "No suggestions for type";
+            }
+        }
+     
+        /// <param name="syntaxType"></param>
+        /// <returns>the common term for the given syntax type</returns>
+        private string SimplifySyntaxTypeString(string syntaxType)
+        {
+            switch (syntaxType)
+            {
+                case "EventFieldDeclarationSyntax":  return "event";
+                case "ConstructorDeclarationSyntax": return "constructor";
+                case "DelegateDeclarationSyntax":    return "delegate";
+                case "IndexerDeclarationSyntax":     return "indexer";
+                case "MethodDeclarationSyntax":      return "method";
+                case "OperatorDeclarationSyntax":    return "operator";
+                case "PropertyDeclarationSyntax":    return "property";
+                default: return "unknown syntax type: " + syntaxType;
+            }
+        }
+
+        private static readonly string[] nonWindowsRuntimeInterfaces = 
+        {
+            "System.Exception",
+            "IAsyncAction",
+            "IAsyncActionWithProgress`1",
+            "IAsyncOperation`1",
+            "IAsyncOperationWithProgress`2",
+        };
+
+        private readonly static string[] InvalidTypes = 
+        { 

```

Comment:
Don't need to address right now, but we need to do a not valid type check rather than an invalid type check.  But can be addressed later on scenarios that we identify if not already handled.

Reply:
Went ahead and changed the name, but not sure I understand the difference in not valid and invalid